aim/aim_lib/db/optimizer.py

```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2018 Cisco Systems
+# All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import functools
+
+from aim.api import resource
+from aim.db import models
+from aim import exceptions
+
+
+class WrongStoreType(exceptions.AimException):
+    message = "Store feature missing for this optimizer call: %(features)s."
+
+
+def requires(requirements):
+    def wrap(func):
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            diff = set(requirements) - set(args[0].store.features)
+            if diff:
+                raise WrongStoreType(features=requirements)
+            else:
+                return func(*args, **kwargs)
+        return inner
+    return wrap
+
+
+@requires(['sql'])
+def get_epg_by_host_names(context, host_names):
```

Comment:
This could just be a function in aim/utils that has a fast implementation for SQL stores and a default brute force one for other stores. That way the AIM library users don't have to worry about special cases.
Also, the name of this function should perhaps be something like 'get_epg_by_static_path_hosts'.